### PR TITLE
core: add timer on pw checks against timing attacks for finding valid usernames (0.x)

### PR DESCRIPTION
**Pull request by mworrell**
### Description

There is a timing difference between checking a password for valid usernames and invalid usernames.

The timing difference can be used to find valid usernames.

This change adds a timer around the pw check to make all pw checks about 300msec.

### Checklist

- [ ] documentation updated
- [ ] tests added
- [x] no BC breaks

### CODE REVIEW

**Comment by mworrell** [comment]
@robvandenbogaard should we also have a check that we don't have multiple pw tries for the same user at the same time?

**Comment by mworrell** [comment]
See also #3098 for a discussion about the timer.

**Review by robvandenbogaard — COMMENTED** [review]
Anti-hammering per username seems sensible, did you mean otherwise one could send many concurrent requests to try multiple passwords?

File: src/models/m_identity.erl [hunk 1]
```diff
@@ -473,11 +480,32 @@ nodash(S) ->
...
 
+
 %% @doc Return the rsc_id with the given username/password.
-%%      If succesful then updates the 'visited' timestamp of the entry.
+%% If succesful then updates the 'visited' timestamp of the entry.
+%% Use a timer to level the time difference between existing and non existing accounts.
 -spec check_username_pw(binary() | string(), binary() | string(), list(), z:context()) ->
             {ok, m_rsc:resource_id()} | {error, term()}.
 check_username_pw(Username, Password, QueryArgs, Context) ->
+    Timeout = ?DEFAULT_PW_CHECK_DURATION + z_ids:number(?DEFAULT_PW_CHECK_VARIANCE),
```
[the remark this answers is not in the document]
Comment (mworrell) [reply src/models/m_identity.erl:490]: Random number generation is a bit in flux in Erlang, that is why we use a wrapper function.

**Comment by mworrell** [comment]
> Anti-hammering per username seems sensible, did you mean otherwise one could send many concurrent requests to try multiple passwords?

Yes, I was thinking about the 300msec period.  The try-counter is only incremented _after_ the check. Alternatively we could increment the try counter _before_ the password check, then we can prevent that someone tries N passwords at once.  In that case we start by assuming that the password check will fail (incrementing the counter), which should be ok, as it only give 1 less check, by allowing N+1 checks we can compensate.

**Review by robvandenbogaard — APPROVED** [review]
I guess the remaining issue(s) (mainly adding _MIN to the .._DURATION constant) are ok to leave as is.

**Pull request merged**